File: android/nativebrik/src/test/java/com/nativebrik/sdk/data/extraction/extraction.kt
package com.nativebrik.sdk.data.extraction

import com.nativebrik.sdk.data.user.UserProperty
import com.nativebrik.sdk.data.user.UserPropertyType
import com.nativebrik.sdk.data.user.getCurrentDate
import com.nativebrik.sdk.schema.ConditionOperator
import com.nativebrik.sdk.schema.ExperimentCondition
import com.nativebrik.sdk.schema.ExperimentConfig
import com.nativebrik.sdk.schema.ExperimentConfigs
import com.nativebrik.sdk.schema.ExperimentVariant
import com.nativebrik.sdk.schema.VariantConfig
import org.junit.Assert
import org.junit.Test

class ExtractionUnitTest {
    @Test
    fun extractComponentId_shouldWork() {
        Assert.assertEquals("hello", extractComponentId(ExperimentVariant(configs = listOf(
            VariantConfig(value = "hello")
        ))))
    }

    @Test
    fun extractExperimentVariant_shouldExtractVariantProbabilistically() {
        val config = ExperimentConfig(
            baseline = ExperimentVariant(id = "1", weight = 1), // 0.25
            variants = listOf(
                ExperimentVariant(id = "2", weight = 1), // 0.5
                ExperimentVariant(id = "3", weight = 1), // 0.75
                ExperimentVariant(id = "4", weight = 1), // 1.0
            )
        )

        Assert.assertEquals("1", extractExperimentVariant(config, 0.24)?.id)
        Assert.assertEquals("2", extractExperimentVariant(config, 0.48)?.id)
        Assert.assertEquals("3", extractExperimentVariant(config, 0.74)?.id)
        Assert.assertEquals("4", extractExperimentVariant(config, 0.99)?.id)
    }

    @Test
    fun extractExperimentVariant_shouldWorkWithoutVariants() {
        val config = ExperimentConfig(
            baseline = ExperimentVariant(id = "1", weight = 1), // 0.25
        )
        Assert.assertEquals("1", extractExperimentVariant(config, 0.24)?.id)
    }

    @Test
    fun extractExperimentConfig_shouldWork() {
        val properties: (seed: Int?) -> List<UserProperty> = {
            emptyList()
        }
        val configs = ExperimentConfigs(
            configs = listOf(
                ExperimentConfig(
                    "1",
                    distribution = emptyList()
                )
            )
        )
        Assert.assertEquals("1", extractExperimentConfig(configs, properties) { _, _ -> true }?.id)
    }

    @Test
    fun extractExperimentConfig_shouldFilterOnlyRunningConfig() {
        val properties: (seed: Int?) -> List<UserProperty> = {
            emptyList()
        }
        val configs = ExperimentConfigs(
            configs = listOf(
                ExperimentConfig(
                    "1",
                    startedAt = getCurrentDate().plusDays(1),
                ),
                ExperimentConfig("2",
                    startedAt = getCurrentDate().minusDays(1),
                    endedAt = getCurrentDate().minusDays(1)
                ),
                ExperimentConfig(
                    "running",
                    startedAt = getCurrentDate().minusDays(1),
                    endedAt = getCurrentDate().plusDays(1),
                )
            )
        )
        Assert.assertEquals("running", extractExperimentConfig(configs, properties) { _, _ -> true }?.id)
    }

    @Test
    fun isInDistributionTarget_shouldBeTrueWhenEmptyDistribution() {
        val properties: List<UserProperty> = emptyList()
        Assert.assertEquals(true, isInDistributionTarget(null, properties))
    }

    @Test
    fun isInDistributionTarget_shouldBeTrueWhenThereCorrectDistribution() {
        val properties: List<UserProperty> = listOf(
            UserProperty(
                name = "name",
                value = "Nativebrik",
                type = UserPropertyType.STRING
            ),
            UserProperty(
                name = "id",
                value = "XXX",
                type = UserPropertyType.STRING
            )
        )
        val distribution: List<ExperimentCondition> = listOf(
            ExperimentCondition(
                property = "name",
                operator = ConditionOperator.Equal.name,
                value = "Nativebrik"
            ),
            ExperimentCondition(
                property = "id",
                operator = ConditionOperator.Equal.name,
                value = "XXX"
            ),
        )
        Assert.assertEquals(true, isInDistributionTarget(distribution, properties))
    }

    @Test
    fun isInDistributionTarget_shouldBeFalseWhenThereIncorrectDistribution() {
        val properties: List<UserProperty> = listOf(
            UserProperty(
                name = "name",
                value = "Nativebrik",
                type = UserPropertyType.STRING
            ),
            UserProperty(
                name = "id",
                value = "XXX",
                type = UserPropertyType.STRING
            )
        )
        val distribution: List<ExperimentCondition> = listOf(
            ExperimentCondition(
                property = "name",
                operator = ConditionOperator.Equal.name,
                value = "Nativebrik"
            ),
            ExperimentCondition(
                property = "id",
                operator = ConditionOperator.Equal.name,
                value = "YYY"
            ),
        )
        Assert.assertEquals(false, isInDistributionTarget(distribution, properties))
    }
}

File: ios/Nativebrik/Nativebrik/data/extraction/extraction.swift
//
//  experiment.swift
//  Nativebrik
//
//  Created by Ryosuke Suzuki on 2023/08/24.
//

import Foundation

func extractComponentId(variant: ExperimentVariant) -> String? {
    guard let configs = variant.configs else {
        return nil
    }
    if configs.count == 0 {
        return nil
    }
    let id = configs.first
    return id?.value
}

/**
    normalizedUsrRnd should be in [0, 1]
 */
func extractExperimentVariant(config: ExperimentConfig, normalizedUsrRnd: Double) -> ExperimentVariant? {
    guard let baseline = config.baseline else {
        return nil
    }

    guard let variants = config.variants else {
        return baseline
    }

    if variants.count == 0 {
        return baseline
    }

    let baselineWeight = baseline.weight ?? 1
    var weights = [baselineWeight]
    var weightSum = baselineWeight
    for variant in variants {
        let variantWeight = variant.weight ?? 1
        weights.append(variantWeight)
        weightSum += variantWeight
    }

    // here is calculation of the picking from the probability.
    // X is selected when p_X(x) >= F_X(x)
    // where F_X(x) := Integral p_X(t)dt, the definition of comulative distribution function
    var comulativeDistributionValue: Double = 0.0
    var selectedVariantIndex: Int = 0
    for (index, weight) in weights.enumerated() {
        let probability: Double = Double(weight) / Double(weightSum)
        comulativeDistributionValue += probability

        if comulativeDistributionValue >= normalizedUsrRnd {
            selectedVariantIndex = index
            break
        }
    }

    if selectedVariantIndex == 0 {
        return baseline
    }

    if variants.count > selectedVariantIndex - 1 {
        return variants[selectedVariantIndex - 1]
    }
    return nil
}

func extractExperimentConfigMatchedToProperties(
    configs: ExperimentConfigs,
    properties: (_ seed: Int) -> [UserProperty],
    isNotInFrequency: (_ experimentId: String, _ frequency: ExperimentFrequency?) -> Bool,
    isMatchedToUserEventFrequencyConditions: (_ conditions: [UserEventFrequencyCondition]?) -> Bool
) -> ExperimentConfig? {
    guard let configs = configs.configs else {
        return nil
    }
    if configs.count == 0 {
        return nil
    }
    let now = getCurrentDate()
    return configs.first { config in
        if let startedAt = config.startedAt {
            if let startedAt = parseDateTime(startedAt) {
                if now.compare(startedAt) == ComparisonResult.orderedAscending {
                    return false
                }
            }
        }
        if let endedAt = config.endedAt {
            if let endedAt = parseDateTime(endedAt) {
                if now.compare(endedAt) == ComparisonResult.orderedDescending {
                    return false
                }
            }
        }

        guard let distribution = config.distribution else {
            return true
        }
        let experimentId = config.id ?? ""
        return isNotInFrequency(experimentId, config.frequency) && isMatchedToUserEventFrequencyConditions(config.eventFrequencyConditions) && isInDistribution(distribution: distribution, properties: properties(config.seed ?? 0))
    }
}

func isInDistribution(distribution: [ExperimentCondition], properties: [UserProperty]) -> Bool {
    let props = Dictionary(uniqueKeysWithValues: properties.map({ property in
        return (property.name, property)
    }))
    let foundNotMatched = distribution.first { condition in
        guard let propKey = condition.property else {
            return true
        }
        guard let conditionValue = condition.value else {
            return true
        }
        guard let op = condition.operator else {
            return true
        }
        guard let prop = props[propKey] else {
            return true
        }
        return !comparePropWithConditionValue(prop: prop, asType: condition.asType, value: conditionValue, op: ConditionOperator(rawValue: op) ?? .Equal)
    }
    return foundNotMatched == nil
}

